rwkv.cpp

```diff
@@ -1277,25 +1505,25 @@ struct rwkv_context * rwkv_clone_context(struct rwkv_context * ctx, const uint32
     return clone;
 }
 
-bool rwkv_gpu_offload_layers(const struct rwkv_context * ctx, const uint32_t n_gpu_layers) {
+bool rwkv_gpu_offload_layers(struct rwkv_context * ctx, const uint32_t n_gpu_layers) {
 #ifdef GGML_USE_CUBLAS
     size_t n_gpu = std::min(n_gpu_layers, ctx->instance->model.header.n_layer);
 
-    size_t gpu_layers = ctx->gpu_layers;
-    size_t vram_total = ctx->vram_total;
+    size_t & gpu_layers = ctx->gpu_layers;
+    size_t & vram_total = ctx->vram_total;
 
-    for (size_t i = 0; i < n_gpu; i++) {
+    for (size_t i = gpu_layers; i < n_gpu; i++) {
         const struct rwkv_layer & layer = ctx->instance->model.layers[i];
 
         // Use cuBLAS only for heavy matrices; other operations are not supported for GPU at the moment
-        ggml_cuda_transform_tensor(layer.att_key);          vram_total += ggml_nbytes(layer.att_key);
-        ggml_cuda_transform_tensor(layer.att_value);        vram_total += ggml_nbytes(layer.att_value);
-        ggml_cuda_transform_tensor(layer.att_receptance);   vram_total += ggml_nbytes(layer.att_receptance);
-        ggml_cuda_transform_tensor(layer.att_output);       vram_total += ggml_nbytes(layer.att_output);
+        ggml_cuda_assign_buffers(layer.att_key);          vram_total += ggml_nbytes(layer.att_key);
+        ggml_cuda_assign_buffers(layer.att_value);        vram_total += ggml_nbytes(layer.att_value);
+        ggml_cuda_assign_buffers(layer.att_receptance);   vram_total += ggml_nbytes(layer.att_receptance);
+        ggml_cuda_assign_buffers(layer.att_output);       vram_total += ggml_nbytes(layer.att_output);
```

Comment:
`ggml_cuda_transform_tensor` has a new argument that we can't figure out how to use. We tried passing it the pointer to the tensor's data like llama.cpp does, but maybe we were reading the wrong portion of llama.cpp's code because `ggml_cuda_transform_tensor` [throws an assert if the tensor isn't already on the GPU](https://github.com/ggerganov/ggml/blob/0a63fc0f6cb1915d1fa5c62c8f0f018d072253c9/src/ggml-cuda.cu#L2489-L2496).

-Emily

Reply:
The intended use is to set the backend to either GPU or GPU_SPLIT (to determine multi GPU behavior) and then call the function with a pointer to the weights.

Reply:
Thank you, that seems to have worked!

-Emily